File: leetcode253/leetcode253_1.cpp
class Solution {
private:
    static bool comp(const vector<int>& x, const vector<int>& y) {
        return x[0] < y[0];
    }
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        // check for the base case. If there is no interval, return 0
        int n = intervals.size();
        if (n == 0) {
            return 0;
        }
        
        // default heap is max heap
        // to change to min heap,  priority_queue <Type, vector<Type>, ComparisonType > min_heap;
        priority_queue<int, vector<int>, greater<int>> pq;
        
        // sort intevals base on start time
        sort(intervals.begin(), intervals.end(), comp);
        
        // add the first meeting
        pq.push(intervals[0][1]);
        
        // iterate the rest intervals
        for (int i = 1; i < n; i++) {
            int start = intervals[i][0], end = intervals[i][1];
            // if the room due to free up the earliest is free, assign that room to this meeting.
            if (pq.top() <= start) {
                pq.pop();
            }
            
            // if a new room is to be assigned, then also we add to the heap,
            // if an old room is allocated, then also we have to add to the heap with updated end tiem
            pq.push(end);
        }
        
        // the size of the heap tells us the minimum rooms required for all the meetings
        return pq.size();
    }
};
// Runtime: 20 ms, faster than 75.25% of C++ online submissions for Meeting Rooms II.
// Memory Usage: 12 MB, less than 100.00% of C++ online submissions for Meeting Rooms II.
// time complexity: O(n log n)
// space complexity: O(n)

File: leetcode253/leetcode253_1.java
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        int n = intervals.length;
        // Check for the base case. If there are no intervals, return 0
        if (n == 0) {
            return 0;
        }

        Integer[] start = new Integer[n];
        Integer[] end = new Integer[n];

        for (int i = 0; i < n; i++) {
            start[i] = intervals[i][0];
            end[i] = intervals[i][1];
        }

        // Sort the intervals by start time
        // 1.
        // Arrays.sort(
        // start,
        // new Comparator<Integer>() {
        // public int compare(Integer a, Integer b) {
        // return a - b;
        // }
        // });
        // 2.
        // Arrays.sort(start, (a, b) -> a - b);
        // 3.
        Arrays.sort(start);

        // Sort the intervals by end time
        // 1.
        // Arrays.sort(
        // end,
        // new Comparator<Integer>() {
        // public int compare(Integer a, Integer b) {
        // return a - b;
        // }
        // });
        // 2.
        // Arrays.sort(end, (a, b) -> a - b);
        // 3.
        Arrays.sort(end);

        // The two pointers in the algorithm: e_ptr and s_ptr.
        int startPointer = 0, endPointer = 0;

        // Variables to keep track of maximum number of rooms used.
        int usedRooms = 0;

        // Iterate over intervals.
        while (startPointer < intervals.length) {

            // If there is a meeting that has ended by the time the meeting at
            // `start_pointer` starts
            if (start[startPointer] >= end[endPointer]) {
                usedRooms -= 1;
                endPointer += 1;
            }

            // We do this irrespective of whether a room frees up or not.
            // If a room got free, then this used_rooms += 1 wouldn't have any effect.
            // used_rooms would
            // remain the same in that case. If no room was free, then this would increase
            // used_rooms
            usedRooms += 1;
            startPointer += 1;

        }

        return usedRooms;
    }
}

// Runtime: 6 ms, faster than 77.89% of Java online submissions for Meeting
// Rooms II.
// Memory Usage: 39.5 MB, less than 31.53% of Java online submissions for
// Meeting Rooms II.
// time complexity: O(n log n)
// space complexity: O(n)